# Ensemble Learning

## Initial Imports

In [10]:
import warnings
warnings.filterwarnings('ignore')

In [11]:
import numpy as np
import pandas as pd
from pathlib import Path
from collections import Counter

In [12]:
from sklearn.metrics import balanced_accuracy_score
from sklearn.metrics import confusion_matrix
from imblearn.metrics import classification_report_imbalanced

## Read the CSV and Perform Basic Data Cleaning

In [13]:
# Load the data
file_path = Path(r'C:\Users\TribThapa\Desktop\Thapa\ResearchFellow\Courses\FinTech_Bootcamp_MonashUni2021\monu-mel-virt-fin-pt-05-2021-u-c\Homework\Week 11\Instructions\Resources\LoanStats_2019Q1.csv')
df = pd.read_csv(file_path)

# Preview the data
df.head()

loan_amnt  int_rate  installment home_ownership  annual_inc  \
0    10500.0    0.1719       375.35           RENT     66000.0   
1    25000.0    0.2000       929.09       MORTGAGE    105000.0   
2    20000.0    0.2000       529.88       MORTGAGE     56000.0   
3    10000.0    0.1640       353.55           RENT     92000.0   
4    22000.0    0.1474       520.39       MORTGAGE     52000.0   

  verification_status   issue_d loan_status pymnt_plan    dti  ...  \
0     Source Verified  Mar-2019    low_risk          n  27.24  ...   
1            Verified  Mar-2019    low_risk          n  20.23  ...   
2            Verified  Mar-2019    low_risk          n  24.26  ...   
3            Verified  Mar-2019    low_risk          n  31.44  ...   
4        Not Verified  Mar-2019    low_risk          n  18.76  ...   

   pct_tl_nvr_dlq  percent_bc_gt_75  pub_rec_bankruptcies  tax_liens  \
0            85.7             100.0                   0.0        0.0   
1            91.2              50.0                   1.0        0.0   
2            66.7              50.0                   0.0        0.0   
3           100.0              50.0                   1.0        0.0   
4           100.0               0.0                   0.0        0.0   

   tot_hi_cred_lim  total_bal_ex_mort total_bc_limit  \
0          65687.0            38199.0         2000.0   
1         271427.0            60641.0        41200.0   
2          60644.0            45684.0         7500.0   
3          99506.0            68784.0        19700.0   
4         219750.0            25919.0        27600.0   

   total_il_high_credit_limit  hardship_flag  debt_settlement_flag  
0                     61987.0              N                     N  
1                     49197.0              N                     N  
2                     43144.0              N                     N  
3                     76506.0              N                     N  
4                     20000.0              N                     N  

[5 rows x 86 columns]

## Split the Data into Training and Testing

In [16]:
# Create our features
X = df.drop(columns=["loan_status"], axis=1)

# Create our target
y = df['loan_status']
y

0        low_risk
1        low_risk
2        low_risk
3        low_risk
4        low_risk
           ...   
68812    low_risk
68813    low_risk
68814    low_risk
68815    low_risk
68816    low_risk
Name: loan_status, Length: 68817, dtype: object

In [17]:
X.head()

loan_amnt  int_rate  installment home_ownership  annual_inc  \
0    10500.0    0.1719       375.35           RENT     66000.0   
1    25000.0    0.2000       929.09       MORTGAGE    105000.0   
2    20000.0    0.2000       529.88       MORTGAGE     56000.0   
3    10000.0    0.1640       353.55           RENT     92000.0   
4    22000.0    0.1474       520.39       MORTGAGE     52000.0   

  verification_status   issue_d pymnt_plan    dti  delinq_2yrs  ...  \
0     Source Verified  Mar-2019          n  27.24          0.0  ...   
1            Verified  Mar-2019          n  20.23          0.0  ...   
2            Verified  Mar-2019          n  24.26          0.0  ...   
3            Verified  Mar-2019          n  31.44          0.0  ...   
4        Not Verified  Mar-2019          n  18.76          0.0  ...   

   pct_tl_nvr_dlq  percent_bc_gt_75  pub_rec_bankruptcies  tax_liens  \
0            85.7             100.0                   0.0        0.0   
1            91.2              50.0                   1.0        0.0   
2            66.7              50.0                   0.0        0.0   
3           100.0              50.0                   1.0        0.0   
4           100.0               0.0                   0.0        0.0   

   tot_hi_cred_lim total_bal_ex_mort  total_bc_limit  \
0          65687.0           38199.0          2000.0   
1         271427.0           60641.0         41200.0   
2          60644.0           45684.0          7500.0   
3          99506.0           68784.0         19700.0   
4         219750.0           25919.0         27600.0   

   total_il_high_credit_limit  hardship_flag  debt_settlement_flag  
0                     61987.0              N                     N  
1                     49197.0              N                     N  
2                     43144.0              N                     N  
3                     76506.0              N                     N  
4                     20000.0              N                     N  

[5 rows x 85 columns]

In [18]:
X.describe()

loan_amnt      int_rate   installment    annual_inc           dti  \
count  68817.000000  68817.000000  68817.000000  6.881700e+04  68817.000000   
mean   16677.594562      0.127718    480.652863  8.821371e+04     21.778153   
std    10277.348590      0.048130    288.062432  1.155800e+05     20.199244   
min     1000.000000      0.060000     30.890000  4.000000e+01      0.000000   
25%     9000.000000      0.088100    265.730000  5.000000e+04     13.890000   
50%    15000.000000      0.118000    404.560000  7.300000e+04     19.760000   
75%    24000.000000      0.155700    648.100000  1.040000e+05     26.660000   
max    40000.000000      0.308400   1676.230000  8.797500e+06    999.000000   

        delinq_2yrs  inq_last_6mths      open_acc       pub_rec  \
count  68817.000000    68817.000000  68817.000000  68817.000000   
mean       0.217766        0.497697     12.587340      0.126030   
std        0.718367        0.758122      6.022869      0.336797   
min        0.000000        0.000000      2.000000      0.000000   
25%        0.000000        0.000000      8.000000      0.000000   
50%        0.000000        0.000000     11.000000      0.000000   
75%        0.000000        1.000000     16.000000      0.000000   
max       18.000000        5.000000     72.000000      4.000000   

           revol_bal  ...  num_tl_90g_dpd_24m  num_tl_op_past_12m  \
count   68817.000000  ...        68817.000000        68817.000000   
mean    17604.142828  ...            0.052138            2.219423   
std     21835.880400  ...            0.390633            1.897432   
min         0.000000  ...            0.000000            0.000000   
25%      6293.000000  ...            0.000000            1.000000   
50%     12068.000000  ...            0.000000            2.000000   
75%     21735.000000  ...            0.000000            3.000000   
max    587191.000000  ...           18.000000           19.000000   

       pct_tl_nvr_dlq  percent_bc_gt_75  pub_rec_bankruptcies  tax_liens  \
count    68817.000000      68817.000000          68817.000000    68817.0   
mean        95.057627         30.626217              0.125972        0.0   
std          8.326426         33.631463              0.336732        0.0   
min         20.000000          0.000000              0.000000        0.0   
25%         93.000000          0.000000              0.000000        0.0   
50%        100.000000         20.000000              0.000000        0.0   
75%        100.000000         50.000000              0.000000        0.0   
max        100.000000        100.000000              4.000000        0.0   

       tot_hi_cred_lim  total_bal_ex_mort  total_bc_limit  \
count     6.881700e+04       6.881700e+04    68817.000000   
mean      2.100332e+05       6.133843e+04    29734.128558   
std       1.928088e+05       5.738798e+04    26795.394232   
min       3.600000e+03       2.350000e+02      100.000000   
25%       6.697700e+04       2.650300e+04    11600.000000   
50%       1.467100e+05       4.535700e+04    22100.000000   
75%       3.036400e+05       7.657000e+04    39300.000000   
max       3.292782e+06       1.295455e+06   509400.000000   

       total_il_high_credit_limit  
count                6.881700e+04  
mean                 5.572240e+04  
std                  5.095845e+04  
min                  1.270000e+02  
25%                  2.288000e+04  
50%                  4.200000e+04  
75%                  7.249900e+04  
max                  1.426964e+06  

[8 rows x 76 columns]

In [19]:
# Check the balance of our target values
df[["loan_status"]].value_counts()

loan_status
low_risk       68470
high_risk        347
dtype: int64

In [23]:
X.columns

Index(['loan_amnt', 'int_rate', 'installment', 'home_ownership', 'annual_inc',
       'verification_status', 'issue_d', 'pymnt_plan', 'dti', 'delinq_2yrs',
       'inq_last_6mths', 'open_acc', 'pub_rec', 'revol_bal', 'total_acc',
       'initial_list_status', 'out_prncp', 'out_prncp_inv', 'total_pymnt',
       'total_pymnt_inv', 'total_rec_prncp', 'total_rec_int',
       'total_rec_late_fee', 'recoveries', 'collection_recovery_fee',
       'last_pymnt_amnt', 'next_pymnt_d', 'collections_12_mths_ex_med',
       'policy_code', 'application_type', 'acc_now_delinq', 'tot_coll_amt',
       'tot_cur_bal', 'open_acc_6m', 'open_act_il', 'open_il_12m',
       'open_il_24m', 'mths_since_rcnt_il', 'total_bal_il', 'il_util',
       'open_rv_12m', 'open_rv_24m', 'max_bal_bc', 'all_util',
       'total_rev_hi_lim', 'inq_fi', 'total_cu_tl', 'inq_last_12m',
       'acc_open_past_24mths', 'avg_cur_bal', 'bc_open_to_buy', 'bc_util',
       'chargeoff_within_12_mths', 'delinq_amnt', 'mo_sin_old_il_acct',

In [34]:
X_binary = pd.get_dummies(X, columns=["home_ownership", "verification_status", "issue_d", "pymnt_plan", "initial_list_status", "next_pymnt_d", "application_type", "hardship_flag", "debt_settlement_flag"])

X_binary.head()

loan_amnt  int_rate  installment  annual_inc    dti  delinq_2yrs  \
0    10500.0    0.1719       375.35     66000.0  27.24          0.0   
1    25000.0    0.2000       929.09    105000.0  20.23          0.0   
2    20000.0    0.2000       529.88     56000.0  24.26          0.0   
3    10000.0    0.1640       353.55     92000.0  31.44          0.0   
4    22000.0    0.1474       520.39     52000.0  18.76          0.0   

   inq_last_6mths  open_acc  pub_rec  revol_bal  ...  issue_d_Mar-2019  \
0             0.0       8.0      0.0     1609.0  ...                 1   
1             0.0      17.0      1.0    18368.0  ...                 1   
2             0.0       8.0      0.0    13247.0  ...                 1   
3             1.0      10.0      1.0    17996.0  ...                 1   
4             1.0      14.0      0.0     9091.0  ...                 1   

   pymnt_plan_0  initial_list_status_f  initial_list_status_w  \
0             1                      0                      1   
1             1                      0                      1   
2             1                      0                      1   
3             1                      0                      1   
4             1                      0                      1   

   next_pymnt_d_Apr-2019  next_pymnt_d_May-2019  application_type_Individual  \
0                      0                      1                            1   
1                      0                      1                            1   
2                      0                      1                            1   
3                      0                      1                            1   
4                      0                      1                            1   

   application_type_Joint App  hardship_flag_N  debt_settlement_flag_N  
0                           0                1                       1  
1                           0                1                       1  
2                           0                1                       1  
3                           0                1                       1  
4                           0                1                       1  

[5 rows x 95 columns]

In [38]:
X_binary.to_csv('Test.csv')

In [35]:
# Split the X and y into X_train, X_test, y_train, y_test
from sklearn.model_selection import train_test_split

X_train, X_test, y_train, y_test = train_test_split(X_binary,
                                                    y,
                                                    random_state=1)

## Data Pre-Processing

Scale the training and testing data using the `StandardScaler` from `sklearn`. Remember that when scaling the data, you only scale the features data (`X_train` and `X_testing`).

In [36]:
# Create the StandardScaler instance
from sklearn.preprocessing import StandardScaler

scaler = StandardScaler()

In [37]:
# Fit the Standard Scaler with the training data. When fitting scaling functions, only train on the training dataset
X_scaler = scaler.fit(X_train)

In [40]:
# Scale the training and testing data
X_train_scaled = X_scaler.transform(X_train)

X_test_scaled = X_scaler.transform(X_test)

## Ensemble Learners

In this section, you will compare two ensemble algorithms to determine which algorithm results in the best performance. You will train a Balanced Random Forest Classifier and an Easy Ensemble classifier . For each algorithm, be sure to complete the folliowing steps:

1. Train the model using the training data. 
2. Calculate the balanced accuracy score from sklearn.metrics.
3. Display the confusion matrix from sklearn.metrics.
4. Generate a classication report using the `imbalanced_classification_report` from imbalanced-learn.
5. For the Balanced Random Forest Classifier only, print the feature importance sorted in descending order (most important feature to least important) along with the feature score

Note: Use a random state of 1 for each algorithm to ensure consistency between tests

### Balanced Random Forest Classifier

In [43]:
# Resample the training data with the BalancedRandomForestClassifier
from sklearn.ensemble import RandomForestClassifier

rf_model = RandomForestClassifier(n_estimators=500,
                                  random_state=42)


rf_model.fit(X_train_scaled, y_train)

RandomForestClassifier(n_estimators=500, random_state=42)

In [44]:
# Calculated the balanced accuracy score
from sklearn.metrics import confusion_matrix, accuracy_score, classification_report

rf_pred = rf_model.predict(X_test_scaled)

rf_acc_score = accuracy_score(y_test, rf_pred)

print(f"The accuracy score for the Balanced Random Forest Classifier is {rf_acc_score}.")

The accuracy score for the Balanced Random Forest Classifier is 0.9960476605637896.


In [45]:
# Display the confusion matrix
rf_cm = confusion_matrix(y_test, rf_pred)

rf_cm_df = pd.DataFrame(rf_cm,
                        index=["Actual False", "Actual True"],
                        columns=["Predicted False", "Predicted True"])

rf_cm_df

Predicted False  Predicted True
Actual False               37              64
Actual True                 4           17100

In [46]:
# Print the imbalanced classification report
from imblearn.metrics import classification_report_imbalanced

print(classification_report_imbalanced(y_test, rf_pred))

                   pre       rec       spe        f1       geo       iba       sup

  high_risk       0.90      0.37      1.00      0.52      0.61      0.34       101
   low_risk       1.00      1.00      0.37      1.00      0.61      0.39     17104

avg / total       1.00      1.00      0.37      1.00      0.61      0.39     17205



In [64]:
# List the features sorted in descending order by feature importance
rf_importances = rf_model.feature_importances_

sorted(zip(rf_model.feature_importances_, X.columns), reverse=True)

[(0.0704577614774606, 'out_prncp'),
 (0.06851571074119261, 'total_acc'),
 (0.06678684949959895, 'total_rec_prncp'),
 (0.059975960224828366, 'initial_list_status'),
 (0.05985277832212354, 'revol_bal'),
 (0.021954993704883744, 'installment'),
 (0.019889661422775016, 'pub_rec'),
 (0.019532515583778935, 'open_acc'),
 (0.016561999715022084, 'acc_open_past_24mths'),
 (0.016287965893022595, 'annual_inc'),
 (0.016109567140325406, 'open_il_12m'),
 (0.015354160817770444, 'inq_last_12m'),
 (0.015110705040483816, 'loan_amnt'),
 (0.015058257081250668, 'delinq_2yrs'),
 (0.01459027319094609, 'all_util'),
 (0.014039682030988047, 'max_bal_bc'),
 (0.01403322788132459, 'num_tl_30dpd'),
 (0.01403092409313799, 'last_pymnt_amnt'),
 (0.01391375341982648, 'num_tl_90g_dpd_24m'),
 (0.013792553383010163, 'tot_coll_amt'),
 (0.013605329463640292, 'total_rev_hi_lim'),
 (0.013065970993287625, 'home_ownership'),
 (0.012639077725431965, 'tot_cur_bal'),
 (0.012623401365315689, 'pct_tl_nvr_dlq'),
 (0.012594603392861226,

### Easy Ensemble Classifier

In [53]:
# Train the Classifier
from imblearn.ensemble import EasyEnsembleClassifier

easy_ens = EasyEnsembleClassifier(n_estimators=10,
                                  random_state=1)

easy_ens.fit(X_train_scaled, y_train)

EasyEnsembleClassifier(random_state=1)

In [59]:
y_test

7034     low_risk
38804    low_risk
65533    low_risk
37736    low_risk
6390     low_risk
           ...   
31854    low_risk
27835    low_risk
3089     low_risk
43343    low_risk
27206    low_risk
Name: loan_status, Length: 17205, dtype: object

In [60]:
# Calculated the balanced accuracy score
easy_ens_pred = easy_ens.predict(X_test_scaled)

from sklearn.metrics import balanced_accuracy_score
easy_ens_bal_acc_score = balanced_accuracy_score(y_test, easy_ens_pred)

print(f"The balanced accuracy score for the Easy Ensemble Classifier is {easy_ens_bal_acc_score}")

The balanced accuracy score for the Easy Ensemble Classifier is 0.9154459266085636


In [61]:
# Display the confusion matrix
easy_ens_cm = confusion_matrix(y_test, easy_ens_pred)

easy_ens_cm_df = pd.DataFrame(easy_ens_cm,
                              index=["Actual False", "Actual True"],
                              columns=["Predicted False", "Predicted True"])

easy_ens_cm_df

Predicted False  Predicted True
Actual False               94               7
Actual True              1707           15397

In [62]:
# Print the imbalanced classification report
print(classification_report_imbalanced(y_test, easy_ens_pred))

                   pre       rec       spe        f1       geo       iba       sup

  high_risk       0.05      0.93      0.90      0.10      0.92      0.84       101
   low_risk       1.00      0.90      0.93      0.95      0.92      0.84     17104

avg / total       0.99      0.90      0.93      0.94      0.92      0.84     17205



### Final Questions

1. Which model had the best balanced accuracy score?

    The Balanced Random Forest Classifier model has a better balanced score (0.9960476605637896) when compared to the Easy Ensemble Classifier (0.9154459266085636) model. 

2. Which model had the best recall score?

    The Easy Ensemble Classifier model has a better recall score (high risk: 0.93 & low risk: 0.90) when compared to the Balanced Random Forest Classifier (high risk: 0.37 & low risk: 1) model.

3. Which model had the best geometric mean score?

    The Easy Ensemble Classifier model has a better geometric score (high risk: 0.92 & low risk: 0.92) when compared to the Balanced Random Forest Classifier (high risk: 0.34 & low risk: 0.39) model.

4. What are the top three features?

    The top 3 features for the Balanced Random Forest Classifier model are: i) out_prncp ii) total_acc, and iii) total_rec_prncp.